File: python/leetcodehard/interleaving_string.py
"""
https://leetcode.com/problems/interleaving-string/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def isInterleave(self, s1, s2, s3):
        if len(s1) + len(s2) != len(s3):
            return False

        rows = len(s1)
        cols = len(s2)
        dp = [[False] * (cols + 1) for _ in range(rows + 1)]
        dp[0][0] = True

        for i in range(1, rows + 1):
            dp[i][0] = dp[i - 1][0] and s1[i - 1] == s3[i - 1]

        for j in range(1, cols + 1):
            dp[0][j] = dp[0][j - 1] and s2[j - 1] == s3[j - 1]

        for i in range(1, rows + 1):
            for j in range(1, cols + 1):
                take_s1 = dp[i - 1][j] and s1[i - 1] == s3[i + j - 1]
                take_s2 = dp[i][j - 1] and s2[j - 1] == s3[i + j - 1]
                dp[i][j] = take_s1 or take_s2

        return dp[rows][cols]


if __name__ == "__main__":
    test_input_1 = ("aabcc", "dbbca", "aadbbcbcac")
    output = Solution().isInterleave(*test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = ("aabcc", "dbbca", "aadbbbaccc")
    output = Solution().isInterleave(*test_input_2)
    Solution().log.info("Output is %s", output)

File: typescript/leetcodehard/interleavingString.ts
/*
https://leetcode.com/problems/interleaving-string/

status - completed
*/

import { logp } from "../utils/logger";

function isInterleave(s1: string, s2: string, s3: string): boolean {
  const m = s1.length;
  const n = s2.length;
  if (m + n !== s3.length) return false;

  const dp: boolean[][] = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(false));
  dp[0][0] = true;

  for (let i = 1; i <= m; i++) {
    dp[i][0] = dp[i - 1][0] && s1[i - 1] === s3[i - 1];
  }
  for (let j = 1; j <= n; j++) {
    dp[0][j] = dp[0][j - 1] && s2[j - 1] === s3[j - 1];
  }

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      const fromS1 = dp[i - 1][j] && s1[i - 1] === s3[i + j - 1];
      const fromS2 = dp[i][j - 1] && s2[j - 1] === s3[i + j - 1];
      dp[i][j] = fromS1 || fromS2;
    }
  }

  return dp[m][n];
}

logp(`input s1="aabcc" s2="dbbca" s3="aadbbcbcac" expected true actual ${isInterleave("aabcc", "dbbca", "aadbbcbcac")}`);
logp(`input s1="aabcc" s2="dbbca" s3="aadbbbaccc" expected false actual ${isInterleave("aabcc", "dbbca", "aadbbbaccc")}`);
logp(`input s1="" s2="" s3="" expected true actual ${isInterleave("", "", "")}`);
